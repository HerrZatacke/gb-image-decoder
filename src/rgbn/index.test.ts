import { describe, expect, test } from 'vitest';
import { hash } from 'ohash';
import { createCanvas } from 'canvas';
import { getRawRGBNImageData } from '.';
import { BlendMode, CanvasCreator, ExportFrameMode, FullRGBNImageCreationParams } from '..';
import tiles16x14 from '../../test/data/tiles/rgbn/16x14';
import tiles20x18 from '../../test/data/tiles/rgbn/20x18';
import { rgbnSoft, rgbnDefault } from '../../test/data/palettes';
import { writeImageFileFromImageData } from '../../test/helpers/writeImageFileFromImageData';

const canvasCreator: CanvasCreator = () => (createCanvas(0, 0) as unknown as HTMLCanvasElement);

describe('RGBN image generation', () => {
  describe('using 16x14 raw data', async () => {
    const fullParams: FullRGBNImageCreationParams = {
      palette: rgbnDefault,
      lockFrame: false,
      tiles: tiles16x14,
      imageStartLine: 0,
      tilesPerLine: 16,
      scaleFactor: 1,
      handleExportFrame: ExportFrameMode.FRAMEMODE_KEEP,
    };

    const {
      data: rawData,
      dimensions,
    } = getRawRGBNImageData(fullParams, canvasCreator);

    test('generates expected dimensions', () => {
      expect(dimensions).toMatchSnapshot();
    });

    test('generates expected hash for raw image data', () => {
      expect(hash(rawData)).toMatchSnapshot();
    });

    test('writes image file without error', async () => {
      await writeImageFileFromImageData(
        'rgbn_16x14.png',
        rawData,
        dimensions,
      );
    });
  });

  describe.each([
    { blendMode: BlendMode.NORMAL },
    { blendMode: BlendMode.NORMAL_S },
    { blendMode: BlendMode.LIGHTEN },
    { blendMode: BlendMode.SCREEN },
    { blendMode: BlendMode.DODGE },
    { blendMode: BlendMode.ADDITION },
    { blendMode: BlendMode.DARKEN },
    { blendMode: BlendMode.MULTIPLY },
    { blendMode: BlendMode.BURN },
    { blendMode: BlendMode.OVERLAY },
    { blendMode: BlendMode.SOFTLIGHT },
    { blendMode: BlendMode.HARDLIGHT },
    { blendMode: BlendMode.DIFFERENCE },
    { blendMode: BlendMode.EXCLUSION },
  ])('with blendmode: $blendMode', async ({ blendMode }) => {

    describe.each([{ lockFrame: true }, { lockFrame: false }])('with lockFrame: $lockFrame', async ({ lockFrame }) => {

      const fullParams: FullRGBNImageCreationParams = {
        palette: {
          ...rgbnSoft,
          blend: blendMode,
        },
        lockFrame,
        tiles: tiles20x18,
        imageStartLine: 2,
        tilesPerLine: 20,
        scaleFactor: 1,
        handleExportFrame: ExportFrameMode.FRAMEMODE_KEEP,
      };

      const {
        data: rawData,
        dimensions,
      } = getRawRGBNImageData(fullParams, canvasCreator);

      test('generates expected dimensions', () => {
        expect(dimensions).toMatchSnapshot();
      });

      test('generates expected hash for raw image data', () => {
        expect(hash(rawData)).toMatchSnapshot();
      });

      test('writes image file without error', async () => {
        await writeImageFileFromImageData(
          `rgbn_20x18_${blendMode}${lockFrame ? '_lockFrame' : ''}.png`,
          rawData,
          dimensions,
        );
      });
    });
  });
});
