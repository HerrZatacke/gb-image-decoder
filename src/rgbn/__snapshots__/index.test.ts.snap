// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RGBN image generation > using 16x14b raw tile data > generates expected dimensions 1`] = `
{
  "height": 112,
  "width": 128,
}
`;

exports[`RGBN image generation > using 16x14b raw tile data > generates expected hash for raw image data 1`] = `"tpQxUeZM2itr8vKKRTgeoCpiUGoGQPUbYoIxDwTBx0U"`;

exports[`RGBN image generation > using 16x14rgb raw tile data > generates expected dimensions 1`] = `
{
  "height": 112,
  "width": 128,
}
`;

exports[`RGBN image generation > using 16x14rgb raw tile data > generates expected hash for raw image data 1`] = `"lRPAnTcp9GgcjczAV-S1b5kn_L8GlmPHR-Wc-0lWjcw"`;

exports[`RGBN image generation > using 16x14rgbn raw tile data > generates expected dimensions 1`] = `
{
  "height": 112,
  "width": 128,
}
`;

exports[`RGBN image generation > using 16x14rgbn raw tile data > generates expected hash for raw image data 1`] = `"ACpGe5Bb5Ty04KOhgthEBSEA2zHzlTRg0txcubASzio"`;

exports[`RGBN image generation > using 16x14rgbx raw tile data > generates expected dimensions 1`] = `
{
  "height": 112,
  "width": 128,
}
`;

exports[`RGBN image generation > using 16x14rgbx raw tile data > generates expected hash for raw image data 1`] = `"lRPAnTcp9GgcjczAV-S1b5kn_L8GlmPHR-Wc-0lWjcw"`;

exports[`RGBN image generation > using 16x14rgn raw tile data > generates expected dimensions 1`] = `
{
  "height": 112,
  "width": 128,
}
`;

exports[`RGBN image generation > using 16x14rgn raw tile data > generates expected hash for raw image data 1`] = `"5ZKeRa6IM0pkizuqQvVgapq7a5nUdrJKavZExIih1bE"`;

exports[`RGBN image generation > using 16x14rn raw tile data > generates expected dimensions 1`] = `
{
  "height": 112,
  "width": 128,
}
`;

exports[`RGBN image generation > using 16x14rn raw tile data > generates expected hash for raw image data 1`] = `"7jyMzLn0fG8wp8yhbWlkFsFGNGksTtAPkZyYYtbx1NE"`;

exports[`RGBN image generation > using 16x14xgb raw tile data > generates expected dimensions 1`] = `
{
  "height": 112,
  "width": 128,
}
`;

exports[`RGBN image generation > using 16x14xgb raw tile data > generates expected hash for raw image data 1`] = `"Ko7e2o7lX8zqXqrxE7T5RRG8C23ComQfBLd0VZ-80m4"`;

exports[`RGBN image generation > using rotation 0 > generates expected dimensions 1`] = `
{
  "height": 112,
  "width": 128,
}
`;

exports[`RGBN image generation > using rotation 0 > generates expected hash for raw image data 1`] = `"lRPAnTcp9GgcjczAV-S1b5kn_L8GlmPHR-Wc-0lWjcw"`;

exports[`RGBN image generation > using rotation 1 > generates expected dimensions 1`] = `
{
  "height": 128,
  "width": 112,
}
`;

exports[`RGBN image generation > using rotation 1 > generates expected hash for raw image data 1`] = `"Okoux0YJfpp0Lc_ZEapa9emW_7ki58_rIZzQ0p-Ud0o"`;

exports[`RGBN image generation > using rotation 2 > generates expected dimensions 1`] = `
{
  "height": 112,
  "width": 128,
}
`;

exports[`RGBN image generation > using rotation 2 > generates expected hash for raw image data 1`] = `"a3jViax3BdSw2ei16Ms4dd5_070I4jAvslTiDx1yg4w"`;

exports[`RGBN image generation > using rotation 3 > generates expected dimensions 1`] = `
{
  "height": 128,
  "width": 112,
}
`;

exports[`RGBN image generation > using rotation 3 > generates expected hash for raw image data 1`] = `"ef3wv2Py6cydA2pEq77WRpR2T5WTOGR-T9QbvGvmvL0"`;

exports[`RGBN image generation > with blendmode: 'addition' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'addition' > with lockFrame: false > generates expected hash for raw image data 1`] = `"WR7V17Yhywwr3S0UY0Vb7duNmVZHcltb6JCsXCkBglU"`;

exports[`RGBN image generation > with blendmode: 'addition' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'addition' > with lockFrame: true > generates expected hash for raw image data 1`] = `"T_uzjlzuSEDmenURu2kyOlqSj8CLSH4q97KO5Haj-SY"`;

exports[`RGBN image generation > with blendmode: 'burn' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'burn' > with lockFrame: false > generates expected hash for raw image data 1`] = `"STSzi3krjvrtr4Cj5w6avIFMV95kUShKDSxCYbO5nE8"`;

exports[`RGBN image generation > with blendmode: 'burn' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'burn' > with lockFrame: true > generates expected hash for raw image data 1`] = `"jUkUAVvM00XrWk34TB09e-n4ATpyMaEdn9sJEUJoMG8"`;

exports[`RGBN image generation > with blendmode: 'darken' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'darken' > with lockFrame: false > generates expected hash for raw image data 1`] = `"RkZKcuRnlvMBlhlJRfOBvGzp1gK14nXNNcKWW8-L5mo"`;

exports[`RGBN image generation > with blendmode: 'darken' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'darken' > with lockFrame: true > generates expected hash for raw image data 1`] = `"ZcwxT8b8vzd6_Ju6JZhJ8YZP8KnH2CNXMa7QkLeJW9g"`;

exports[`RGBN image generation > with blendmode: 'difference' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'difference' > with lockFrame: false > generates expected hash for raw image data 1`] = `"GWayRCBLKMgPcNSAQDeJS4ohR9Ejr-elbtg0LO0wbzQ"`;

exports[`RGBN image generation > with blendmode: 'difference' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'difference' > with lockFrame: true > generates expected hash for raw image data 1`] = `"F5phTVPOlbfmK6kr8wEFKYLg3aUH53gboNPCU5wE8DQ"`;

exports[`RGBN image generation > with blendmode: 'dodge' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'dodge' > with lockFrame: false > generates expected hash for raw image data 1`] = `"2JLhRViaPNRqjdrYhsm_9xOuVtZtojfGTaH3oe1jRQo"`;

exports[`RGBN image generation > with blendmode: 'dodge' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'dodge' > with lockFrame: true > generates expected hash for raw image data 1`] = `"NNv_MPoXLEuWNchRBkzgisVB1BtUHI1j2vELlhshmog"`;

exports[`RGBN image generation > with blendmode: 'exclusion' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'exclusion' > with lockFrame: false > generates expected hash for raw image data 1`] = `"Kkl78XEGTOATtZsmVWGNXYR4fx6kOO8-sPB_Q8vvjnk"`;

exports[`RGBN image generation > with blendmode: 'exclusion' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'exclusion' > with lockFrame: true > generates expected hash for raw image data 1`] = `"BdFkrwUWWtS3MsPEAdjOV3Js2f2cCpOg2OR1Slu5hcY"`;

exports[`RGBN image generation > with blendmode: 'hardlight' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'hardlight' > with lockFrame: false > generates expected hash for raw image data 1`] = `"Vo_TouSW_liWvDK9QndY1yixIwtCuVr04hd-sqMWk38"`;

exports[`RGBN image generation > with blendmode: 'hardlight' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'hardlight' > with lockFrame: true > generates expected hash for raw image data 1`] = `"PW2z9lThLwLwnZox8qdeBNv6448slyWIvIOhrWmum0s"`;

exports[`RGBN image generation > with blendmode: 'lighten' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'lighten' > with lockFrame: false > generates expected hash for raw image data 1`] = `"RhXuVENQNuW3Z6kgW5NVn76MhZ7HzNzqDiVBkjM4J-U"`;

exports[`RGBN image generation > with blendmode: 'lighten' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'lighten' > with lockFrame: true > generates expected hash for raw image data 1`] = `"qvhkfpWW9pgWuPdWXGB-ROCUHjBZmPA9CK4kzTsUCe0"`;

exports[`RGBN image generation > with blendmode: 'multiply' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'multiply' > with lockFrame: false > generates expected hash for raw image data 1`] = `"57XbUtmbZcfKRgA8dCNor-cvRB1vtxrulauBnhDy5B8"`;

exports[`RGBN image generation > with blendmode: 'multiply' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'multiply' > with lockFrame: true > generates expected hash for raw image data 1`] = `"fwLtegSCtGBAawPsjx344bg8QHZa90QBJYdr8S_n6Ok"`;

exports[`RGBN image generation > with blendmode: 'normal' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'normal' > with lockFrame: false > generates expected hash for raw image data 1`] = `"Ywmcvo9VzkvbM3ThuNpe6Ij9Y2u9rJWsVXF1cqQtDy0"`;

exports[`RGBN image generation > with blendmode: 'normal' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'normal' > with lockFrame: true > generates expected hash for raw image data 1`] = `"N89wwmI7ub1RQDAnKl5gHSCu8a8IRtoouN5IkuNfgVI"`;

exports[`RGBN image generation > with blendmode: 'normal_s' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'normal_s' > with lockFrame: false > generates expected hash for raw image data 1`] = `"YyQJI00ctSeeeYOQVCcgJzDu2RBu-CJdJ0uVg_BXJXo"`;

exports[`RGBN image generation > with blendmode: 'normal_s' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'normal_s' > with lockFrame: true > generates expected hash for raw image data 1`] = `"oWe1KcZqG4W-uRA_yYZYxhdA7pgFWgXwOwdTdwyj_Sk"`;

exports[`RGBN image generation > with blendmode: 'overlay' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'overlay' > with lockFrame: false > generates expected hash for raw image data 1`] = `"glnY5AYbDY7WM9tFxobfw68dGGeQ84J0yrYlNxQ6zGg"`;

exports[`RGBN image generation > with blendmode: 'overlay' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'overlay' > with lockFrame: true > generates expected hash for raw image data 1`] = `"99K0vMs9QYwh9LFslbbwwiRMmw8eALXKXodIDjm708s"`;

exports[`RGBN image generation > with blendmode: 'screen' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'screen' > with lockFrame: false > generates expected hash for raw image data 1`] = `"RU1IZtbY0tsPnS1EY58yA30I_-0m2UnDrld-AgsJp2s"`;

exports[`RGBN image generation > with blendmode: 'screen' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'screen' > with lockFrame: true > generates expected hash for raw image data 1`] = `"iYzAqvHQqDC7Ncjk6kfe3D5aXNnhx-08BK40qpV3smY"`;

exports[`RGBN image generation > with blendmode: 'softlight' > with lockFrame: false > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'softlight' > with lockFrame: false > generates expected hash for raw image data 1`] = `"nl0O7ZFWw_KUe6nr3avwRsILYkAria9t7SGg6Ku275w"`;

exports[`RGBN image generation > with blendmode: 'softlight' > with lockFrame: true > generates expected dimensions 1`] = `
{
  "height": 144,
  "width": 160,
}
`;

exports[`RGBN image generation > with blendmode: 'softlight' > with lockFrame: true > generates expected hash for raw image data 1`] = `"MJzOPfQAwfWb4tcNbYggKmuHccaluLXvaOsnHAUrvOQ"`;
